server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java

```diff
@@ -392,74 +394,114 @@ private static OutputStream blobOutputStream(Path file) throws IOException {
     }
 
     @Override
-    @SuppressForbidden(reason = "write to channel that we have open for locking purposes already directly")
+    public void getRegister(OperationPurpose purpose, String key, ActionListener<OptionalBytesReference> listener) {
+        // no lock to acquire here, we are emulating the lack of read/read and read/write contention in cloud repositories
+        ActionListener.completeWith(
+            listener,
+            () -> doUncontendedCompareAndExchangeRegister(path.resolve(key), BytesArray.EMPTY, BytesArray.EMPTY)
+        );
+    }
+
+    @Override
     public void compareAndExchangeRegister(
         OperationPurpose purpose,
         String key,
         BytesReference expected,
         BytesReference updated,
         ActionListener<OptionalBytesReference> listener
     ) {
-        ActionListener.completeWith(listener, () -> {
-            BlobContainerUtils.ensureValidRegisterContent(updated);
-            try (LockedFileChannel lockedFileChannel = LockedFileChannel.open(path.resolve(key))) {
-                final FileChannel fileChannel = lockedFileChannel.fileChannel();
-                final ByteBuffer readBuf = ByteBuffer.allocate(BlobContainerUtils.MAX_REGISTER_CONTENT_LENGTH);
-                while (readBuf.remaining() > 0) {
-                    if (fileChannel.read(readBuf) == -1) {
-                        break;
-                    }
+        ActionListener.completeWith(listener, () -> doCompareAndExchangeRegister(path.resolve(key), expected, updated));
+    }
+
+    private static final KeyedLock<Path> writeMutexes = new KeyedLock<>();
+
+    private static OptionalBytesReference doCompareAndExchangeRegister(Path registerPath, BytesReference expected, BytesReference updated)
+        throws IOException {
+        // Emulate write/write contention as might happen in cloud repositories, at least for the case where the writers are all in this
+        // JVM (e.g. for an ESIntegTestCase).
+        try (var mutex = writeMutexes.tryAcquire(registerPath)) {
+            if (mutex == null) {
+                return OptionalBytesReference.MISSING;
+            } else {
+                try {
+                    return doUncontendedCompareAndExchangeRegister(registerPath, expected, updated);
+                } finally {
+                    mutex.close();
```

Comment:
Is this not handled by the try-with-resource above?

Reply:
d'oh yes it is